Add osd.window-switcher.style-classic.item.active.{border,bg}.color

tokyo4j opened the pull request:
For post-0.9.2.

These theme options configures the border/background of selected window item in the `classic` style window siwtcher. Previously, the OSD text color (default: black) was used as the border and background was transparent.

Now the default theme is updated to use #589bda as the border color and #c7e2fc as the background color, which are the same as the `thumbnail` style window switcher.

Screenshot of the new default look:

<img width="1278" height="718" alt="20251001_14h35m54s_grim" src="https://github.com/user-attachments/assets/0d4164d1-e994-445f-b864-7a54d48236f6" />

I'll appreciate feedbacks. If the default colors are not preferable, I think we can update the default colors of `thumbnail` style window switcher for consistency before 0.9.2 release.

Consolatis commented:
This is `Numix` with `#c7e2fc50`:
<img width="1280" height="720" alt="20251001_18h24m47s_grim" src="https://github.com/user-attachments/assets/b5ec9e8b-9150-4b88-a499-26d5b3479462" />

And this is the default theme:
<img width="1280" height="720" alt="20251001_18h24m12s_grim" src="https://github.com/user-attachments/assets/6164f6e2-0fc2-416c-a0b5-1c6988ed44f0" />

tokyo4j commented:
That's a good point. To fix it, I updated the default colors like this:

- `osd.window-switcher.style-classic.item.active.border` inherits `osd.label.text.color` as before.
- `osd.window-switcher.style-classic.item.active.bg` inherits `osd.window-switcher.style-classic.item.active.border` with 15% opacity.

To keep subpixel rendering for fonts, `switcher_classic_theme->item_active_bg_color` stores the blending result of `osd.bg.color` and 15% opacity `osd.window-switcher.style-classic.item.active.border`.

I personally prefer the blue border and light-blue background like shown in the description, but I think we should make it inherit `osd.label.text.color` as before, since some theme may have light-blue osd text color which makes the text unreadable.

Also, I updated the default color of `thumbnail` style window switcher.

Consolatis commented on src/theme.c:
```diff
@@ -1721,6 +1757,26 @@ post_processing(struct theme *theme)
 		memcpy(theme->osd_border_color, theme->osd_label_text_color,
 			sizeof(theme->osd_border_color));
 	}
+	if (switcher_classic_theme->item_active_border_color[0] == FLT_MIN) {
+		memcpy(switcher_classic_theme->item_active_border_color,
+			theme->osd_label_text_color,
+			sizeof(theme->osd_label_text_color));
+	}
+	if (switcher_classic_theme->item_active_bg_color[0] == FLT_MIN) {
+		blend_color_with_bg(switcher_classic_theme->item_active_bg_color,
+			switcher_classic_theme->item_active_border_color, 0.15,
+			theme->osd_bg_color);
+	}
```
Thoughts about this variant? It reduces the bright border in dark themes and seems preferable to me but that is just a personal preference.
```c
	if (switcher_classic_theme->item_active_border_color[0] == FLT_MIN) {
		blend_color_with_bg(switcher_classic_theme->item_active_border_color,
			theme->osd_label_text_color, 0.5,
			theme->osd_bg_color);
	}
	if (switcher_classic_theme->item_active_bg_color[0] == FLT_MIN) {
		blend_color_with_bg(switcher_classic_theme->item_active_bg_color,
			theme->osd_label_text_color, 0.15,
			theme->osd_bg_color);
	}
```

tokyo4j replied:
Looks good to me. I agree that the pure-black border in the default theme and the pure-white border in `Numix` theme are too prominent and I like your 50% gray border.

tokyo4j replied:
Done.

tokyo4j commented:
Now the window switcher looks like this:

||`style=classic`|`style=thumbnail`|
|-|-|-|
|default|<img width="1281" height="721" alt="20251003_14h03m03s_grim" src="https://github.com/user-attachments/assets/7c296a93-4115-41c9-858e-1b70561949be" /> | <img width="1278" height="720" alt="20251003_14h03m18s_grim" src="https://github.com/user-attachments/assets/2d03d96a-a470-4aed-aaea-1cb04597064e" /> |
|Numix| <img width="1275" height="719" alt="20251003_14h03m31s_grim" src="https://github.com/user-attachments/assets/d5bf1ca1-0e6b-4e3f-9992-c9353a6690c1" /> | <img width="1278" height="721" alt="20251003_14h03m40s_grim" src="https://github.com/user-attachments/assets/dcc91e2e-7278-442f-a016-5dbfb8cc5203" /> |

johanmalm commented:
I think that inheriting `osd.label.text.color` with alpha as you've done it makes a lot of sense to avoid out-of-the-box surprises against existing themes.

I'm confused by the values in `docs/themerc`. How can `r`, `g` and `b` be different if the default is `#000000 50%` and `#000000 15%`? I think I'm right in saying that the defaults comes from falling back on `osd.label.text.color` which in turns falls back on `window.active.label.text.color` which is `#000000` by default.


```
osd.window-switcher.style-thumbnail.item.active.border.color: #706f6d
osd.window-switcher.style-thumbnail.item.active.bg.color: #bfbcba
```

LGTM other that. Have tested - works well.

I wasn't aware of `osd.active.label.bg`. It does not appear to be in the wiki, but interestingly it exists in the Clearlooks-3.4 `themerc` file. I don't think that Openbox's theme spec for `osd` is really sufficient for us, so current trajectory makes sense (i.e. ignore `osd.active.label.bg`).

tokyo4j commented:
> I'm confused by the values in docs/themerc. How can r, g and b be different if the default is #000000 50% and #000000 15%? I think I'm right in saying that the defaults comes from falling back on osd.label.text.color which in turns falls back on window.active.label.text.color which is #000000 by default.

`#000000 50%` and `#706f6d` are basically the same. The downside of describing `#000000 50%` in `themerc` is that it disables subpixel antialiasing for the window switcher fields/titles, which slightly degrades UX when users copy `docs/themerc` to their theme.

Also, do you have any thoughts on https://github.com/labwc/labwc/pull/3123#issuecomment-3364251460? I prefer merging only the first commit for now, because I think the changes in `osd-classic.c` is bigger and harder to follow compared to the changes in `theme.c`.

tokyo4j commented:
Opened #3127 for 0.9.2.

> Just wanting to point out that we can also easily delay the release, both options are fine to me.

Yes, let's do that if there's still room for improvement in #3127 or we're busy and it takes several days to merge. The release date is much less important than labwc's stability and our health :smile:
